File: src/main/java/org/hildan/minecraft/mining/optimizer/patterns/DiggingPattern.java
package org.hildan.minecraft.mining.optimizer.patterns;

import org.hildan.minecraft.mining.optimizer.chunks.Sample;

import java.util.Set;

/**
 * Represents a way to dig into the stone in 3 dimensions.
 */
public interface DiggingPattern {

    /**
     * Gets the width of this pattern. This dimension is related to the X coordinate.
     *
     * @return the width of this pattern
     */
    int getWidth();

    /**
     * Gets the height of this pattern. This dimension is related to the Y coordinate.
     *
     * @return the height of this pattern
     */
    int getHeight();

    /**
     * Gets the length of this pattern. This dimension is related to the Z coordinate.
     *
     * @return the length of this pattern
     */
    int getLength();

    /**
     * Gives the coordinates where the player has to enter a sample to start digging this pattern. Multiple accesses may
     * be returned, meaning the player has to enter each of them independently to dig this pattern.
     * <p>
     * The accesses' positions in the sample depend on the position of the pattern within the sample.
     *
     * @param x
     *         the X position of this pattern within the sample
     * @param y
     *         the Y position of this pattern within the sample
     * @return the set of accesses at the given pattern position
     */
    Set<Access> getAccesses(int x, int y);

    /**
     * Digs this pattern into the given sample. The pattern is repeated as many times as necessary in every direction,
     * starting from the point (0,0,0).
     */
    void digInto(Sample sample);
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/patterns/DiggingPattern.kt
package org.hildan.minecraft.mining.optimizer.patterns

import org.hildan.minecraft.mining.optimizer.blocks.Sample

/**
 * Represents a way to dig into the stone in 3 dimensions.
 */
interface DiggingPattern {
    /**
     * Digs this pattern into the given [sample]. This method must take care of stopping at the edge of the given sample.
     */
    fun digInto(sample: Sample)

    /**
     * Digs this pattern into the given [sample], and then digs every visible ore recursively until no more ore is
     * directly visible. This is supposed to represent what a real user will do while digging the pattern.
     */
    fun digAndFollowOres(sample: Sample) {
        digInto(sample)
        sample.digVisibleOresRecursively()
    }
}
